### PR TITLE
feat(pipeline): Application pipeline auto binding definition

### PR DESCRIPTION
**Pull request by littlejiancc**
#### What this PR does / why we need it:
 Application pipeline auto binding definition

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMzIxXSwiYXNzaWduZWUiOlsiMTAwMTI2MSJdfQ%3D%3D&id=325515&iterationID=1321&pId=323719&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |    Application pipeline auto binding definition          |
| 🇨🇳 中文    |    应用流水线自动绑定流水线定义          |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: internal/tools/pipeline/services/pipelinesvc/create_v2.go [hunk 1]
```diff
@@ -414,4 +414,4 @@ func constructToCreateCronRequest(p *spec.Pipeline, cronStartFrom *time.Time, co
 		PipelineDefinitionID: p.PipelineDefinitionID,
 	}
 	return createReq
-}
```
Comment (sfwn) [inline internal/tools/pipeline/services/pipelinesvc/create_v2.go:417]: do not remove end empty newline.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/project_pipeline.go [hunk 2]
```diff
@@ -74,3 +74,7 @@ func GetRuleCategoryKeyMap() map[string]PipelineCategory {
 	}
 	return m
 }
+
+func MakeRemote(app *ApplicationDTO) string {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply apistructs/project_pipeline.go]: done

File: internal/pkg/mock/source_mock.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+package mock
+
+import (
+	"context"
+
+	"github.com/erda-project/erda-proto-go/core/pipeline/source/pb"
+)
+
+type SourceMock struct {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply internal/pkg/mock/source_mock.go:23]: done

File: internal/tools/pipeline/services/pipelinesvc/create_v2.go [hunk 4]
```diff
@@ -415,3 +424,116 @@ func constructToCreateCronRequest(p *spec.Pipeline, cronStartFrom *time.Time, co
 	}
 	return createReq
 }
+
+func (s *PipelineSvc) GetDefinitionID(req *apistructs.PipelineCreateRequestV2) (string, error) {
+	if req.DefinitionID != "" {
+		return req.DefinitionID, nil
+	}
+	if req.PipelineSource != apistructs.PipelineSourceDice {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply internal/tools/pipeline/services/pipelinesvc/create_v2.go]: done

File: internal/tools/pipeline/services/pipelinesvc/create_v2.go [hunk 5]
```diff
@@ -415,3 +424,116 @@ func constructToCreateCronRequest(p *spec.Pipeline, cronStartFrom *time.Time, co
...
+		return "", nil
+	}
+	return s.CreatePipelineSourceAndDefinition(context.Background(), req)
+}
+
+func (s *PipelineSvc) CreatePipelineSourceAndDefinition(ctx context.Context, req *apistructs.PipelineCreateRequestV2) (string, error) {
+	const (
+		sourceType = "erda"
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply internal/tools/pipeline/services/pipelinesvc/create_v2.go]: put to dop

File: internal/tools/pipeline/services/pipelinesvc/create_v2.go [hunk 6]
```diff
@@ -415,3 +424,116 @@ func constructToCreateCronRequest(p *spec.Pipeline, cronStartFrom *time.Time, co
 	}
 	return createReq
 }
+
+func (s *PipelineSvc) GetDefinitionID(req *apistructs.PipelineCreateRequestV2) (string, error) {
```
[the remark this answers is not in the document]
Comment (littlejiancc) [reply internal/tools/pipeline/services/pipelinesvc/create_v2.go]: done

File: internal/tools/pipeline/services/pipelinesvc/create_v2.go (hunk 1 above)
Comment (littlejiancc) [reply internal/tools/pipeline/services/pipelinesvc/create_v2.go]: done

File: internal/apps/dop/providers/projectpipeline/provider.go [hunk 7]
```diff
@@ -85,6 +92,120 @@ func (p *provider) Init(ctx servicehub.Context) error {
...
+		if err := p.AddDefinitionToCronIfNeed(ctx); err != nil {
+			p.Log.Errorf("failed to add definition to cron, err: %v", err.Error())
+		}
+	}()
+	return nil
+}
+
+func (p *provider) AddDefinitionToCronIfNeed(ctx context.Context) error {
+	cronResp, err := p.PipelineCron.CronPaging(ctx, &cronpb.CronPagingRequest{
+		Sources: []string{apistructs.PipelineSourceDice.String()},
```
Comment (sfwn) [inline internal/apps/dop/providers/projectpipeline/provider.go:106]: filter by pipeline_definition_id

File: internal/apps/dop/providers/projectpipeline/provider.go [hunk 8]
```diff
@@ -85,6 +92,120 @@ func (p *provider) Init(ctx servicehub.Context) error {
...
+		GetAll:  true,
+	})
+	if err != nil {
+		return err
+	}
+	corns := cronsFilterIn(cronResp.Data, func(cron *pipelinepb.Cron) bool {
```
Comment (sfwn) [inline internal/apps/dop/providers/projectpipeline/provider.go:112]: typo `corns`

File: internal/pkg/mock/source_mock.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfwn) [reply internal/pkg/mock/source_mock.go:23]: nothing changed

File: internal/tools/pipeline/provider.go [hunk 9]
```diff
@@ -83,6 +85,9 @@ type provider struct {
 	Resource     resource.Interface
 	ActionAgent  actionagent.Interface
 	Permission   permission.Interface
+
+	PipelineSource     sourcepb.SourceServiceServer `autowired:"erda.core.pipeline.source.SourceService" required:"true"`
```
Comment (sfwn) [inline internal/tools/pipeline/provider.go:89]: no need

File: internal/tools/pipeline/services/pipelinesvc/define.go [hunk 10]
```diff
@@ -69,6 +71,9 @@ type PipelineSvc struct {
 	actionMgr    actionmgr.Interface
 	mysql        mysqlxorm.Interface
 	resource     resource.Interface
+
+	pipelineSource     sourcepb.SourceServiceServer
```
Comment (sfwn) [inline internal/tools/pipeline/services/pipelinesvc/define.go:75]: no need

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/tools/pipeline/provider.go (hunk 9 above)
Comment (littlejiancc) [reply internal/tools/pipeline/provider.go:89]: done

File: internal/tools/pipeline/services/pipelinesvc/define.go (hunk 10 above)
Comment (littlejiancc) [reply internal/tools/pipeline/services/pipelinesvc/define.go:75]: done

File: internal/apps/dop/providers/projectpipeline/provider.go (hunk 8 above)
Comment (littlejiancc) [reply internal/apps/dop/providers/projectpipeline/provider.go:112]: done

File: internal/apps/dop/providers/projectpipeline/provider.go (hunk 7 above)
Comment (littlejiancc) [reply internal/apps/dop/providers/projectpipeline/provider.go:106]: done

File: internal/tools/pipeline/services/pipelinesvc/define.go (hunk 10 above)
Comment (littlejiancc) [reply internal/tools/pipeline/services/pipelinesvc/define.go:75]: done

File: internal/tools/pipeline/provider.go (hunk 9 above)
Comment (littlejiancc) [reply internal/tools/pipeline/provider.go:89]: done

**Review by sfwn — APPROVED** [review]

**Pull request merged**